lib/prest/client.rb

```diff
@@ -0,0 +1,74 @@
+# frozen_string_literal: true
+
+module Prest
+  # Main wrapper class for Prest. To use the gem, call:
+  # Prest::Client.new('https://base_uri.com', { headers: { 'Authorization' => 'Bearer token' }})
+  class Client < BasicObject
+    SUPPORTED_HTTP_VERBS = %i[get post put patch delete get! post! put! patch! delete!].freeze
+
+    def initialize(base_uri, options = {})
+      @base_uri = base_uri
+      @options = options
+      @fragments = []
+      @query_params = {}
+    end
+
+    def method_missing(method, *args, **kwargs)
+      if SUPPORTED_HTTP_VERBS.include?(method)
+        if method.to_s.end_with?('!')
+          execute_query!(method[0..-2], **kwargs)
+        else
+          execute_query(method, **kwargs)
+        end
+      else
+        chain_fragment(method.to_s, *args, **kwargs)
+        self
+      end
+    end
+
+    def respond_to_missing?(_, _)
+      true
+    end
+
+    private
+
+    def execute_query(http_method, body: {})
+      res = ::HTTParty.send(http_method, build_url, headers: headers, body: body)
+      ::Prest::Response.new(res.code, res.parsed_response, res.headers)
+    rescue ::HTTParty::ResponseError => e
+      ::Kernel.raise ::Prest::Error, e.message
+    end
+
+    def execute_query!(*args, **kwargs)
+      res = execute_query(*args, **kwargs)
+      ::Kernel.raise ::Prest::Error, res.body.to_json unless res.successful?
+
+      res
+    end
+
+    def chain_fragment(fragment_name, *args, **kwargs)
+      arguments = args.join('/')
+      parsed_args = arguments.empty? ? '' : "/#{arguments}"
+      @query_params.merge!(kwargs)
+      @fragments << "#{fragment_name.gsub("__", "-")}#{parsed_args}"
+    end
+
+    def headers
+      tmp_headers = @options[:headers] || {}
+      tmp_headers.merge!('Content-Type' => 'application/json', 'Accept' => 'application/json') if @options[:json]
```

Comment:
Would be nice if we could do this at the execute method also.

We may only want a specific requests to use this:
```ruby
Client.users.get({ json: true })
```

WDYT?

Reply:
this interferes with how the body is passed to the requests. Remember that the body is passed in the post/put/patch methods.